LSP: `workspace/diagnostic/refresh` from tsgo fails to deserialize — `invalid type: map, expected unit`

### Summary

With the [`tsgo`](https://zed.dev/extensions/tsgo) extension as the TypeScript language server, Zed periodically logs:

```
ERROR [lsp] error deserializing workspace/diagnostic/refresh request: Error("invalid type: map, expected unit", line: 0, column: 0)
```

Per the LSP spec, `workspace/diagnostic/refresh` takes **no params**, so Zed appears to deserialize it into `()`. The server seems to send `params` as an (empty?) object, and the strict `unit` deserialization then fails. If so, the fix would be tolerating `params: {}` / `params: null` on this request.

### Environment

| | |
|---|---|
| Zed | 1.13.2 (latest stable at time of filing) |
| Extension | `tsgo` ("TypeScript Language Server") 1.1.0 |
| Language server | `@typescript/typescript-darwin-arm64/lib/tsc --lsp --stdio`, TypeScript **7.0.2** |
| OS | macOS 24.6.0, arm64 |
| Project | ~840 `.ts`/`.tsx` files, Vite + React, single `tsconfig.json` with `paths` |

Settings:

```jsonc
"TypeScript": { "language_servers": ["typescript-ls", "!vtsls"] }
```

(`vtsls` is explicitly disabled — see "Not vtsls" below.)

### Frequency

Bursty, mostly during initial project indexing: **10 occurrences within ~20s** of opening the project on a cold start, then silent. On a subsequent warm start it fired **once**. So it is intermittent rather than continuous.

Immediately before the burst, this also appears:

```
WARN [project::lsp_store] skipping diagnostics update, no worktree found for path "/users/sanghee/dev/<project>/tsconfig.json"
```

Note the lowercased `/users/...` in that path, which may or may not be related.

### What I could not reproduce

I drove the same server binary directly over stdio (outside Zed) and could **not** get it to emit `workspace/diagnostic/refresh`, across three attempts:

1. `initialize` with `workspace.diagnostics.refreshSupport: true` → `initialized` → `textDocument/didOpen` on a project file, 8s wait.
2. Same, plus `textDocument.diagnostic.{dynamicRegistration,relatedDocumentSupport}` and `publishDiagnostics.relatedInformation`, 40s wait, logging every server→client method.
3. Same as (2), plus `workspace/didChangeWatchedFiles` on `tsconfig.json` (the server registers `typescript-config-watch-id` for config watching) and a real `utime` touch of that file, 20s wait.

In all three the server sent only `window/logMessage` and `client/registerCapability` — never a diagnostic refresh. So I could not capture the offending payload, and I cannot say from the outside whether the server is out of spec or the client is over-strict. Raising it here since the deserialization is client-side; happy to move it to `zed-extensions/tsgo` if that is the better home.

### Impact

Not confirmed. My expectation is that Zed simply ignores the failed refresh request, so diagnostics would not be re-pulled when the server asks — meaning they could go stale until the next edit triggers a pull. I have not observed a concrete stale-diagnostic case, so I do not want to overstate this; the log noise is the only certain symptom.

### Not vtsls

Worth separating out, since I initially suspected the fallback server. With `"language_servers": ["typescript-ls", "vtsls"]`, Zed started **vtsls with its working directory set to the tsgo binary's lib folder** rather than the project:

```
INFO [lsp] starting language server process. binary path: ".../node/bin/node",
  working directory: "~/Library/Application Support/Zed/extensions/work/tsgo/node_modules/@typescript/typescript-darwin-arm64/lib",
  args: [".../@vtsls/language-server/bin/vtsls.js", "--stdio"]
INFO [project::lsp_store] language server 4: unregistered workspace/DidChangeWatchedFiles capability with id program files for /dev/null/inferred watcher
```

It fell into an inferred project (`/dev/null/inferred`, no tsconfig) and reported phantom `Cannot find module 'src/...'` and `Cannot find module './x.module.css'` errors on files that `tsc -p tsconfig.json` compiles cleanly. Disabling it with `!vtsls` removed those false diagnostics — but **the `workspace/diagnostic/refresh` error persists without vtsls**, so the two are independent. I can file the vtsls cwd behaviour separately if useful.

--- a/crates/lsp/src/lsp.rs
+++ b/crates/lsp/src/lsp.rs
@@ -236,6 +236,22 @@ fn is_unit<T: 'static>(_: &T) -> bool {
     TypeId::of::<T>() == TypeId::of::<()>()
 }
 
+fn deserialize_params<T: DeserializeOwned + 'static>(params: Value) -> serde_json::Result<T> {
+    if TypeId::of::<T>() == TypeId::of::<()>() {
+        serde_json::from_value(Value::Null)
+    } else {
+        serde_json::from_value(params)
+    }
+}
+
+fn deserialize_result<T: DeserializeOwned + 'static>(result: &str) -> serde_json::Result<T> {
+    if TypeId::of::<T>() == TypeId::of::<()>() {
+        serde_json::from_str("null")
+    } else {
+        serde_json::from_str(result)
+    }
+}
+
 /// Language server protocol RPC request message.
 ///
 /// [LSP Specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#requestMessage)
@@ -1209,12 +1225,12 @@ impl LanguageServer {
     fn on_custom_notification<Params, F>(&self, method: &'static str, mut f: F) -> Subscription
     where
         F: 'static + FnMut(Params, &mut AsyncApp) + Send,
-        Params: DeserializeOwned,
+        Params: DeserializeOwned + 'static,
     {
         let prev_handler = self.notification_handlers.lock().insert(
             method,
             Box::new(move |_, params, cx| {
-                if let Some(params) = serde_json::from_value(params).log_err() {
+                if let Some(params) = deserialize_params(params).log_err() {
                     f(params, cx);
                 }
             }),
@@ -1243,7 +1259,7 @@ impl LanguageServer {
             method,
             Box::new(move |id, params, cx| {
                 if let Some(id) = id {
-                    match serde_json::from_value(params) {
+                    match deserialize_params(params) {
                         Ok(params) => {
                             let response = f(params, cx);
                             let task = cx.foreground_executor().spawn({
@@ -1467,7 +1483,7 @@ impl LanguageServer {
                         executor
                             .spawn(async move {
                                 let response = match result {
-                                    Ok(response) => match serde_json::from_str(&response) {
+                                    Ok(response) => match deserialize_result(&response) {
                                         Ok(deserialized) => Ok(deserialized),
                                         Err(error) => {
                                             log::error!("failed to deserialize response from language server: {}. response from language server: {:?}", error, response);
@@ -2239,6 +2255,86 @@ mod tests {
         drop(subscription);
         assert!(detached_payload_handle.upgrade().is_none());
         assert!(retained_payload_handle.upgrade().is_none());
+    }
+
+    #[gpui::test]
+    async fn test_unit_params_request_with_empty_object_params(cx: &mut TestAppContext) {
+        cx.update(|cx| {
+            release_channel::init(semver::Version::new(0, 0, 0), cx);
+        });
+        let (server, fake) = FakeLanguageServer::new(
+            LanguageServerId(0),
+            LanguageServerBinary {
+                path: "path/to/language-server".into(),
+                arguments: Vec::new(),
+                env: None,
+            },
+            "the-lsp".to_string(),
+            Default::default(),
+            &mut cx.to_async(),
+        );
+
+        enum DiagnosticRefreshWithRawParams {}
+
+        impl request::Request for DiagnosticRefreshWithRawParams {
+            type Params = Value;
+            type Result = ();
+            const METHOD: &'static str = request::WorkspaceDiagnosticRefresh::METHOD;
+        }
+
+        let (refresh_tx, refresh_rx) = async_channel::unbounded();
+        server
+            .on_request::<request::WorkspaceDiagnosticRefresh, _, _>(move |(), _| {
+                let refresh_tx = refresh_tx.clone();
+                async move {
+                    refresh_tx.try_send(()).unwrap();
+                    Ok(())
+                }
+            })
+            .detach();
+
+        for params in [serde_json::json!({}), Value::Null] {
+            let response = fake
+                .request::<DiagnosticRefreshWithRawParams>(params, DEFAULT_LSP_REQUEST_TIMEOUT)
+                .await;
+            assert_eq!(response.into_response().unwrap(), ());
+            assert_eq!(refresh_rx.recv().await, Ok(()));
+        }
+    }
+
+    #[gpui::test]
+    async fn test_unit_result_response_with_empty_object(cx: &mut TestAppContext) {
+        cx.update(|cx| {
+            release_channel::init(semver::Version::new(0, 0, 0), cx);
+        });
+        let (server, fake) = FakeLanguageServer::new(
+            LanguageServerId(0),
+            LanguageServerBinary {
+                path: "path/to/language-server".into(),
+                arguments: Vec::new(),
+                env: None,
+            },
+            "the-lsp".to_string(),
+            Default::default(),
+            &mut cx.to_async(),
+        );
+
+        enum ShutdownWithRawResult {}
+
+        impl request::Request for ShutdownWithRawResult {
+            type Params = Value;
+            type Result = Value;
+            const METHOD: &'static str = request::Shutdown::METHOD;
+        }
+
+        fake.set_request_handler::<ShutdownWithRawResult, _, _>(|_, _| async move {
+            Ok(serde_json::json!({}))
+        });
+
+        let response = server
+            .request::<request::Shutdown>((), DEFAULT_LSP_REQUEST_TIMEOUT)
+            .await;
+        assert_eq!(response.into_response().unwrap(), ());
     }
 
     #[gpui::test]

--- a/crates/lsp/src/lsp.rs
+++ b/crates/lsp/src/lsp.rs
@@ -272,9 +272,9 @@ where
 struct AnyResponse<'a> {
     jsonrpc: &'a str,
     id: RequestId,
-    #[serde(default)]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
     error: Option<Error>,
-    #[serde(borrow)]
+    #[serde(borrow, skip_serializing_if = "Option::is_none")]
     result: Option<&'a RawValue>,
 }
 
@@ -2362,6 +2362,24 @@ mod tests {
             .expect("message with string id should be parsed");
         let expected_id = RequestId::Int(2);
         assert_eq!(notification.id, Some(expected_id));
+    }
+
+    #[test]
+    fn test_serialize_error_response_has_no_result() {
+        let response = AnyResponse {
+            jsonrpc: JSON_RPC_VERSION,
+            id: RequestId::Int(0),
+            error: Some(Error {
+                code: -32601,
+                message: "Unrecognized method".to_string(),
+                data: None,
+            }),
+            result: None,
+        };
+        assert_eq!(
+            serde_json::to_string(&response).unwrap(),
+            "{\"jsonrpc\":\"2.0\",\"id\":0,\"error\":{\"code\":-32601,\"message\":\"Unrecognized method\",\"data\":null}}"
+        );
     }
 
     #[test]

